File: tclk-bridge/bin/tclk-bridge.mjs
#!/usr/bin/env node
// SPDX-License-Identifier: Apache-2.0
// Thin JSON CLI over @flop-labs/tclk. Does NOT reimplement the state machine.
// Never custodies keys or funds. PaperRail is demo-only (settles nothing of value).
// Preimages/secrets are returned only in-memory to the caller; never written to disk by this CLI.

import { createInterface } from "node:readline";
import {
  makeOffer,
  makeAccept,
  generateHashLock,
  hashLockFromPreimage,
  openContract,
  applyFrame,
  encodeFrame,
  decodeFrame,
  tryDecodeFrame,
  isTclkLine,
  verifyHashPreimage,
  verifySecret,
  validateFrame,
  validateDeadlines,
  dealRoom,
  capabilityToken,
  parseCapabilityToken,
  OFFER_ROOM,
  TCLK_VERSION,
  TCLK_PREFIX,
  TCLK_TERMINAL_STATUSES,
  lockTerms,
  PaperRail,
  MemoryNoteStore,
  MemoryRail,
  stateNote,
  stateNoteValue,
  parseStateNoteValue,
  tclkStatusToA2A,
} from "@flop-labs/tclk";

/** In-process paper note store for demo paper_* commands only. */
const paperNotes = new MemoryNoteStore();
const paperRail = new PaperRail(paperNotes);
const memoryRail = new MemoryRail("memory-demo");

function ok(result) {
  return { ok: true, result };
}

function fail(error, code = "error") {
  return { ok: false, error: String(error), code };
}

/** Strip secret from contract state before returning durable views. */
function publicState(state) {
  if (!state || typeof state !== "object") return state;
  const { secret: _secret, ...rest } = state;
  return rest;
}

/** Redact secret from reveal frames for transcript persistence. */
function redactFrame(frame) {
  if (!frame || typeof frame !== "object") return frame;
  if (frame.type === "reveal" && "secret" in frame) {
    const { secret: _s, ...rest } = frame;
    return { ...rest, secret: "[REDACTED]" };
  }
  return frame;
}

async function dispatch(cmd) {
  const action = cmd?.action;
  if (!action) throw new Error("missing action");

  switch (action) {
    case "ping":
      return ok({
        version: TCLK_VERSION,
        prefix: TCLK_PREFIX,
        offerRoom: OFFER_ROOM,
        note: "bridge wraps @flop-labs/tclk; PaperRail settles nothing of value",
      });

    case "makeOffer":
      return ok(makeOffer(cmd.fields));

    case "makeAccept":
      return ok(makeAccept(cmd.offer, cmd.accept));

    case "generateHashLock": {
      // Caller must hold preimage; bridge does not persist it.
      const lock = generateHashLock();
      return ok({
        hash: lock.hash,
        preimage: lock.preimage,
        warning: "preimage is ephemeral — never log or persist server-side",
      });
    }

    case "hashLockFromPreimage":
      return ok(hashLockFromPreimage(cmd.preimage));

    case "openContract":
      return ok(publicState(openContract(cmd.offer)));

    case "applyFrame": {
      const nowMs = typeof cmd.nowMs === "number" ? cmd.nowMs : Date.now();
      const step = applyFrame(cmd.state, cmd.frame, nowMs);
      return ok({
        ...step,
        state: publicState(step.state),
        // Do not echo revealed secret back into durable state.
        revealed: cmd.frame?.type === "reveal" ? Boolean(cmd.frame?.secret) : false,
      });
    }

    case "foldTranscript": {
      // Fold offer + frames through the real machine. Secrets in reveal frames
      // are used for verification then dropped from returned state.
      const nowMs = typeof cmd.nowMs === "number" ? cmd.nowMs : Date.now();
      let state = openContract(cmd.offer);
      const steps = [];
      for (const frame of cmd.frames || []) {
        const step = applyFrame(state, frame, nowMs);
        state = step.state;
        steps.push({
          ok: step.ok,
          reason: step.reason,
          status: state.status,
          frameType: frame?.type,
        });
        if (!step.ok && cmd.stopOnReject) break;
      }
      return ok({ state: publicState(state), steps, terminal: TCLK_TERMINAL_STATUSES.has(state.status) });
    }

    case "encodeFrame":
      return ok({ line: encodeFrame(cmd.frame) });

    case "decodeFrame":
      return ok(redactFrame(decodeFrame(cmd.text)));

    case "tryDecodeFrame": {
      const frame = tryDecodeFrame(cmd.text);
      return ok(frame ? redactFrame(frame) : null);
    }

    case "isTclkLine":
      return ok({ isTclk: isTclkLine(cmd.text) });

    case "validateFrame":
      return ok(redactFrame(validateFrame(cmd.value)));

    case "verifyHashPreimage":
      return ok({ valid: verifyHashPreimage(cmd.hash, cmd.preimage) });

    case "verifySecret":
      return ok({
        valid: verifySecret(cmd.lock, cmd.statement, cmd.secret),
      });

    case "validateDeadlines":
      return ok({
        valid: validateDeadlines(
          cmd.offer,
          cmd.nowMs ?? Date.now(),
          cmd.minClaimWindowMs,
          cmd.minRefundGapMs,
        ),
      });

    case "dealRoom":
      return ok({ room: dealRoom(cmd.contract) });

    case "capabilityToken":
      return ok({ token: capabilityToken(cmd.rails || []) });

    case "parseCapabilityToken":
      return ok({ rails: parseCapabilityToken(cmd.note) });

    case "stateNote":
      return ok(stateNote(cmd.contract));

    case "stateNoteValue":
      return ok({ value: stateNoteValue(cmd.status, cmd.railRef) });

    case "parseStateNoteValue":
      return ok(parseStateNoteValue(cmd.value));

    case "lockTerms":
      return ok(lockTerms(cmd.state));

    case "tclkStatusToA2A":
      return ok({ a2a: tclkStatusToA2A(cmd.status) });

    // --- PaperRail demo (settles NOTHING of value) ---
    case "paperLock": {
      const ref = await paperRail.lock(cmd.terms);
      return ok({
        ref,
        rail: "paper",
        settlement: "unverified",
        warning: "PaperRail records lifecycle only — no funds held",
      });
    }

    case "paperVerifyLock":
      return ok({
        verified: await paperRail.verifyLock(cmd.terms, cmd.ref),
        settlement: "paper-demo",
        warning: "Paper verification is not economic settlement",
      });

    case "paperClaim": {
      await paperRail.claim(cmd.ref, cmd.secret);
      return ok({
        claimed: true,
        settlement: "unverified",
        warning: "PaperRail claim is choreography only",
      });
    }

    case "paperRefund": {
      await paperRail.refund(cmd.ref);
      return ok({
        refunded: true,
        settlement: "unverified",
        warning: "PaperRail refund is choreography only",
      });
    }

    case "paperRead": {
      const record = await paperRail.read(cmd.ref);
      // Strip secret from paper record before returning for persistence paths.
      if (record && record.secret) {
        const { secret: _s, ...rest } = record;
        return ok({
          record: { ...rest, secret: "[REDACTED]" },
          settlement: "unverified",
        });
      }
      return ok({ record, settlement: "unverified" });
    }

    case "memoryLock": {
      const ref = await memoryRail.lock(cmd.terms);
      return ok({
        ref,
        rail: "memory-demo",
        settlement: "unverified",
        warning: "MemoryRail is in-process demo only — no funds",
      });
    }

    default:
      throw new Error(`unknown action: ${action}`);
  }
}

async function main() {
  const rl = createInterface({ input: process.stdin, crlfDelay: Infinity });
  let buf = "";
  for await (const line of rl) {
    buf += line;
  }
  buf = buf.trim();
  if (!buf) {
    process.stdout.write(JSON.stringify(fail("empty stdin", "bad_request")) + "\n");
    process.exit(2);
  }
  let cmd;
  try {
    cmd = JSON.parse(buf);
  } catch (e) {
    process.stdout.write(JSON.stringify(fail(`invalid JSON: ${e.message}`, "bad_request")) + "\n");
    process.exit(2);
  }
  try {
    const out = await dispatch(cmd);
    process.stdout.write(JSON.stringify(out) + "\n");
    process.exit(out.ok ? 0 : 1);
  } catch (e) {
    process.stdout.write(JSON.stringify(fail(e.message || e)) + "\n");
    process.exit(1);
  }
}

main();

File: src/tar/tclk_bridge.py
"""Subprocess bridge to the real @flop-labs/tclk package."""
from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path
from typing import Any

_REPO = Path(__file__).resolve().parents[2]
_DEFAULT_BRIDGE = _REPO / "tclk-bridge" / "bin" / "tclk-bridge.mjs"
_SECRET_KEYS = frozenset({"preimage", "secret", "private_key", "privateKey"})


class TclkBridgeError(RuntimeError):
    def __init__(self, message: str, *, code: str = "bridge_error", detail: Any = None):
        super().__init__(message)
        self.code = code
        self.detail = detail


def _bridge_path() -> Path:
    override = os.environ.get("TCLK_BRIDGE_PATH")
    if override:
        return Path(override)
    return _DEFAULT_BRIDGE


def _redact_for_log(obj: Any) -> Any:
    if isinstance(obj, dict):
        out: dict[str, Any] = {}
        for k, v in obj.items():
            if k in _SECRET_KEYS:
                out[k] = "[REDACTED]"
            else:
                out[k] = _redact_for_log(v)
        return out
    if isinstance(obj, list):
        return [_redact_for_log(x) for x in obj]
    return obj


def call_tclk(command: dict[str, Any], *, timeout: float = 30.0) -> Any:
    bridge = _bridge_path()
    if not bridge.is_file():
        raise TclkBridgeError(f"tclk bridge not found at {bridge}", code="bridge_missing")
    payload = json.dumps(command, separators=(",", ":"), ensure_ascii=True)
    try:
        proc = subprocess.run(
            ["node", str(bridge)],
            input=payload,
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
            cwd=str(bridge.parent.parent),
        )
    except FileNotFoundError as exc:
        raise TclkBridgeError("node runtime not found", code="node_missing") from exc
    except Exception as exc:
        if exc.__class__.__name__ == "TimeoutExpired":
            raise TclkBridgeError("tclk bridge timed out", code="bridge_timeout") from exc
        raise
    raw = (proc.stdout or "").strip()
    if not raw:
        err = (proc.stderr or "").strip() or f"exit {proc.returncode}"
        raise TclkBridgeError(f"empty bridge stdout: {err}", code="bridge_empty")
    try:
        body = json.loads(raw.splitlines()[-1])
    except json.JSONDecodeError as exc:
        raise TclkBridgeError("bridge returned non-JSON", code="bridge_bad_json") from exc
    if not body.get("ok"):
        raise TclkBridgeError(
            str(body.get("error") or "bridge error"),
            code=str(body.get("code") or "bridge_error"),
            detail=_redact_for_log(body),
        )
    return body.get("result")


def ping() -> dict[str, Any]:
    return call_tclk({"action": "ping"})


def make_offer(fields: dict[str, Any]) -> dict[str, Any]:
    return call_tclk({"action": "makeOffer", "fields": fields})


def make_accept(offer: dict[str, Any], accept: dict[str, Any]) -> dict[str, Any]:
    return call_tclk({"action": "makeAccept", "offer": offer, "accept": accept})


def generate_hash_lock() -> dict[str, Any]:
    """Mint a hash lock. Caller must treat preimage as ephemeral."""
    return call_tclk({"action": "generateHashLock"})


def open_contract(offer: dict[str, Any]) -> dict[str, Any]:
    return call_tclk({"action": "openContract", "offer": offer})


def apply_frame(
    state: dict[str, Any], frame: dict[str, Any], now_ms: int | None = None
) -> dict[str, Any]:
    cmd: dict[str, Any] = {"action": "applyFrame", "state": state, "frame": frame}
    if now_ms is not None:
        cmd["nowMs"] = now_ms
    return call_tclk(cmd)


def fold_transcript(
    offer: dict[str, Any],
    frames: list[dict[str, Any]],
    *,
    now_ms: int | None = None,
    stop_on_reject: bool = False,
) -> dict[str, Any]:
    cmd: dict[str, Any] = {
        "action": "foldTranscript",
        "offer": offer,
        "frames": frames,
        "stopOnReject": stop_on_reject,
    }
    if now_ms is not None:
        cmd["nowMs"] = now_ms
    return call_tclk(cmd)


def encode_frame(frame: dict[str, Any]) -> str:
    return call_tclk({"action": "encodeFrame", "frame": frame})["line"]


def try_decode_frame(text: str) -> dict[str, Any] | None:
    return call_tclk({"action": "tryDecodeFrame", "text": text})


def is_tclk_line(text: str) -> bool:
    return bool(call_tclk({"action": "isTclkLine", "text": text})["isTclk"])


def deal_room(contract: str) -> str:
    return call_tclk({"action": "dealRoom", "contract": contract})["room"]


def capability_token(rails: list[str]) -> str:
    return call_tclk({"action": "capabilityToken", "rails": rails})["token"]


def verify_hash_preimage(hash_hex: str, preimage: str) -> bool:
    return bool(
        call_tclk(
            {"action": "verifyHashPreimage", "hash": hash_hex, "preimage": preimage}
        )["valid"]
    )


def paper_lock(terms: dict[str, Any]) -> dict[str, Any]:
    """Demo only — PaperRail holds no value."""
    return call_tclk({"action": "paperLock", "terms": terms})


def paper_verify_lock(terms: dict[str, Any], ref: str) -> dict[str, Any]:
    return call_tclk({"action": "paperVerifyLock", "terms": terms, "ref": ref})


def lock_terms(state: dict[str, Any]) -> dict[str, Any]:
    return call_tclk({"action": "lockTerms", "state": state})
